Source/ARTRealtimePresence.m

```diff
@@ -780,85 +778,130 @@ - (void)reenterInternalMembers {
     return _internalMembers;
 }
 
-- (BOOL)add:(ARTPresenceMessage *)message {
-    ARTPresenceMessage *latest = [_members objectForKey:message.memberKey];
-    if ([message isNewerThan:latest]) {
-        ARTPresenceMessage *messageCopy = [message copy];
+- (void)processMember:(ARTPresenceMessage *)message {
+    ARTPresenceMessage *messageCopy = [message copy];
+    // Internal member
+    if ([message.connectionId isEqualToString:self.connectionId]) {
         switch (message.action) {
             case ARTPresenceEnter:
             case ARTPresenceUpdate:
-                messageCopy.action = ARTPresencePresent;
-                // intentional fallthrough
             case ARTPresencePresent:
-                [self internalAdd:messageCopy];
+                messageCopy.action = ARTPresencePresent;
+                [self addInternalMember:messageCopy];
                 break;
             case ARTPresenceLeave:
-                [self internalRemove:messageCopy];
+                if (!message.isSynthesized) {
+                    [self removeInternalMember:messageCopy];
+                }
                 break;
             default:
                 break;
         }
-        return YES;
     }
-    ARTLogDebug(_logger, @"Presence member \"%@\" with action %@ has been ignored", message.memberKey, ARTPresenceActionToStr(message.action));
-    latest.syncSessionId = _syncSessionId;
-    return NO;
-}
+    
+    BOOL memberUpdated = false;
+    switch (message.action) {
+        case ARTPresenceEnter:
+        case ARTPresenceUpdate:
+        case ARTPresencePresent:
+            messageCopy.action = ARTPresencePresent;
+            memberUpdated = [self addMember:messageCopy];
+            break;
+        case ARTPresenceLeave:
+            if (self.syncInProgress) {
+                messageCopy.action = ARTPresenceAbsent; // RTP2f
+                memberUpdated = [self addMember:messageCopy];
+            } else {
+                memberUpdated = [self removeMember:messageCopy];
+            }
+            break;
+        default:
+            break;
+    }
 
-- (void)internalAdd:(ARTPresenceMessage *)message {
-    [self internalAdd:message withSessionId:_syncSessionId];
+    if (memberUpdated) {
+        [self broadcast:message];
+    }
+    else {
+        ARTLogDebug(_logger, @"Presence member \"%@\" with action %@ has been ignored", message.memberKey, ARTPresenceActionToStr(message.action));
+    }
+    [_beforeSyncMembers removeObjectForKey:message.memberKey]; // RTP19
```

Comment:
As per [RTP19](https://sdk.ably.com/builds/ably/specification/main/features/#RTP19), `_beforeSyncMembers` should only track list of members added or updated and not `leave` event. You can see the same being implemented across SDK's
[ably-go](https://github.com/ably/ably-go/blob/dd8236fc301eb078803e83efee7615f3ba133341/ably/realtime_presence.go#L318)
[ably-java](https://github.com/ably/ably-java/blob/f706160e09d639bf482474d13898933481e455b9/lib/src/main/java/io/ably/lib/realtime/Presence.java#L1056)
[ably-dotnet](https://github.com/ably/ably-dotnet/blob/f7b18f4902e42e90697e7fd35021025574edf650/src/IO.Ably.Shared/Realtime/PresenceMap.cs#L87)
None of them track/update `beforeSyncMembers` for `LEAVE` event.

Reply:
https://github.com/ably/ably-cocoa/pull/1891/commits/00b04a0b020c6515d1d99dfcd8237f4d0430abe9